### ISSUE TITLE
Should Converters ever touch data files?

### ISSUE THREAD

**Issue by luiztauffer** [catalystneuro/to-nwb-template#2]
Take for example the template experiment: it has interfaces for Ecephys, Ophys and Behavior, but labs might have metadata living outside the data files interfaced by the main DataInterfaces. For example Subject metadata or even metadata that concerns the data types of these main interfaces, but its not reached by them. I can see two solutions to solve that:

1. LabConverter fetches its DataInterfaces info with `get_input_schema()`, `get_metadata_schema()` and `get_metadata()` and further expand these methods to fetch lab/experiment specific info. Example with `get_input_schema()`:
```python
class MyLabNWBConverter(NWBConverter):

    # Modular data interfaces
    data_interface_classes = {
        'EcephysDataInterface': EcephysDataInterface,
        'OphysDataInterface': OphysDataInterface,
        'BehaviorDataInterface': BehaviorDataInterface
    }

    @classmethod
    def get_input_schema(cls):
        """Get input schema from data interfaces and from lab-specific source"""
        # Get data-specific input schemas from data interfaces
        input_schema = super(TemplateNWBConverter, cls).get_metadata()

        # Get lab-specific input schema
        with pkg_resources.open_text(schema, 'input_schema_labspecific.json') as f:
            input_schema_labspecific = json.load(f)

        # Update data-specific input schema with lab-specific schema
        input_schema['properties'] = dict_deep_update(
            input_schema['properties'],
            input_schema_labspecific['properties']
        )

        return input_schema
```

2. Any lab / experiment that requires specific info fetching would also require a lab-specific DataInterface, that would do the data handling. This would actually leave the Converters as simple thin wrappers for executing lists of DataInterfaces. Example:
```python
class MyLabNWBConverter(NWBConverter):

    # Modular data interfaces
    data_interface_classes = {
        'EcephysDataInterface': EcephysDataInterface,
        'OphysDataInterface': OphysDataInterface,
        'BehaviorDataInterface': BehaviorDataInterface,
		'MyLabDataInterface': MyLabDataInterface
    }
```

**Comment by bendichter** [comment]
I think you could use method 1 for small additions and 2 for large additions or additions that you think might generalize to other converters.

**Comment by CodyCBakerPhD** [comment]
I currently use option #1 for all my metadata handling, but I can see the points now for trying to generalize some of the fields in handing them off to the individual interfaces. Reconciling this completely with [Issue #5 ](https://github.com/catalystneuro/to-nwb-template/issues/5) will take a bit more thought, but for the moment I do think I'll begin adhering to what Ben said where the `ExperimenterNWBConverter` class's `get_metadata` a) retrieves most of it's metadata from the `DataInterface` classes but b) appends/updates a few small custom fields that are not included there.

I always like to think about these things contextually with reference to a working example. As ugly as [this code I wrote for the yuta conversion](https://github.com/catalystneuro/buzsaki-lab-to-nwb/blob/master/buzsaki_lab_to_nwb/yuta_code/yutanwbconverter.py#L262-L289) is, it does properly hand off the all the metadata to the `RecordingExtractor `and others for writing purposes. Now, all fields except the `theta_reference` stuff ought to be generalizable across all Neuroscope formats and will be included in the `NeuroscopeRecordingDataInterface` as a `DataInterface` specific `get_metadata()` retrieval in an upcoming PR.

Specific to that conversion though, and in line with Ben's comment, the small addition of the `theta_reference` stuff (highly specific to Yuta, at least in terms of how exactly it is retrieved in the data) would still belong inside the `YutaNWBConverter` `get_metadata()`, appended after the primary retrieval from the future `NeuroscopeRecordingDataInterface`.

**Comment by CodyCBakerPhD** [comment]
TL;DR version,

Yes, NWBConverters can touch data files but are only encouraged to do so for synchronization-level tasks or other highly customized and specific things that generic interface-specific options cannot handle.

But mostly you should let the data interfaces do their job on their respective data sources.
